manual/src/refman/extensions/tail_mod_cons.etex

```diff
@@ -0,0 +1,548 @@
+the [@tail_mod_cons] attribute, or mark this call with
+the [@tailcall false] attribute to make its non-tailness explicit.
+\end{verbatim}
+
+Again the fix is to specialized "append_flatten" as well:
+\begin{verbatim}
+let[@tail_mod_cons] rec flatten = function
+| [] -> []
+| xs :: xss ->
+  append_flatten xs xss
+
+and[@tail_mod_cons] append_flatten xs xss =
+  match xs with
+  | [] -> flatten xss
+  | x :: xs -> x :: append_flatten xs xss
+\end{verbatim}
+
+Non-recursive functions can also be marked "[\@tail_mod_cons]"; this is
+typically useful for local bindings to recursive functions.
+
+Incorrect version:
+\begin{verbatim}
+let[@tail_mod_cons] rec map_vars f exp =
+  let self exp = map_vars f exp in
+  match exp with
+  | Var v -> Var (f v)
+  | Let ((v, def), body) ->
+    Let ((f v, self def), (self[@tailcall]) body)
+                           ^^^^^^^^^^^^^^^^^^^^^
+Warning 51 [wrong-tailcall-expectation]: expected tailcal
+\end{verbatim}
+
+Recommend fix:
+\begin{verbatim}
+let[@tail_mod_cons] rec map_vars f exp =
+  let[@tail_mod_cons] self exp = map_vars f exp in
+  match exp with
+  | Var v -> Var (f v)
+  | Let ((v, def), body) ->
+    Let ((f v, self def), (self[@tailcall]) body)
+\end{verbatim}
+
+In other cases, there is no benefit in making the called function
+tail-mod-cons, or it is not possible, for example it is a function
+parameter (the transformation only works with calls to
+known functions).
+
+For example, consider a substitution function on binary trees:
+\begin{verbatim}
+type 'a tree = Leaf of 'a | Node of 'a tree * 'a tree
+
+let[@tail_mod_cons] rec bind (f : 'a -> 'a tree) (t : 'a tree) : 'a tree =
+  match t with
+  | Leaf v -> f v
+  | Node (left, right) -> Node (bind f left, (bind[@tailcall]) f right)
+
+|             f v
+              ^^^
+Warning 72 [tmc-breaks-tailcall]: This call is in tail-modulo-cons position in a TMC function,
+but the function called is not itself specialized for TMC,
+so the call will not be transformed into a tail call.
+Please either mark the called function with
+the [@tail_mod_cons] attribute, or mark this call with
+the [@tailcall false] attribute to make its non-tailness explicit.
+\end{verbatim}
+
+Here "f" is a function parameter, not a direct call, and the current
+implementation is strictly first-order, it does not support
+tail-mod-cons arguments. In this case, the user should indicate that
+they realize this call to "f v" is not, in fact, in tail position, by
+using "(f[\@tailcall false]) v".
+
+\begin{verbatim}
+type 'a tree = Leaf of 'a | Node of 'a tree * 'a tree
+
+let[@tail_mod_cons] rec bind (f : 'a -> 'a tree) (t : 'a tree) : 'a tree =
+  match t with
+  | Leaf v -> (f[@tailcall false]) v
+  | Node (left, right) -> Node (bind f left, (bind[@tailcall]) f right)
+\end{verbatim}
+
+\subsection{ss:details}{Details on the transformation}
+
+To use this advanced feature, it helps to have an idea of what the
+program transformation does.
```

Comment:
```suggestion
program transformation is.
```
?

Reply:
This phrase looks to me too weak compared to the rest of the text: it is always better to use tools which you understand.  to rewrite this introductory sentence slightly differently, like

"To discuss the general idea and implementation details of TMC transformation let us recall our 'map' example:"

Reply:
> it is always better to use tools which you understand

Yes and no; in general we strive for abstractions that are robust, users don't have to know what is inside to get a good mental model of how to use it. In my experience, TMC is not as good an abstraction, and it helps to know some implementation details to think about its usage. (Not all implementation details matter, but the idea of destination-passing-style certainly does, even though it does not seem apparent in the "interface" of the feature.)

Writing this as a reply to your comment did give me a better idea of something to write:

> To understand well how to use this advanced feature, it helps to be aware that the function transformation produces a specialized function in destination-passing-style.

Thanks!